core/src/main/scala/quasar/plugin/postgres/CsvSink.scala

```diff
@@ -0,0 +1,213 @@
+          .through(copyToTable(tbl, cols))
+          .compile
+          .drain
+
+      ensureTable =
+        writeMode match {
+          case WriteMode.Create =>
+            createTable(tbl, cols)
+
+          case WriteMode.Replace =>
+            dropTableIfExists(tbl) >> createTable(tbl, cols)
+        }
+
+      _ <- (ensureTable >> PHC.pgGetCopyAPI(doCopy)).transact(xa) handleErrorWith { t =>
+        error[F](s"COPY to '${tbl}' produced unexpected error: ${t.getMessage}", t) >>
+          AE.raiseError(t)
+      }
+
+      endAt <- timer.clock.monotonic(MILLISECONDS)
+      tbytes <- totalBytes.get
+
+      _ <- debug[F](s"SUCCESS: COPY ${tbytes} bytes to '${tbl}' in ${endAt - startAt} ms")
+
+    } yield ()
+  }
+
+  ////
+
+  private val logHandler: LogHandler =
+    LogHandler {
+      case Success(q, _, e, p) =>
+        log.debug(s"SUCCESS: `$q` in ${(e + p).toMillis}ms (${e.toMillis} ms exec, ${p.toMillis} ms proc)")
+
+      case ExecFailure(q, _, e, t) =>
+        log.debug(s"EXECUTION_FAILURE: `$q` after ${e.toMillis} ms, detail: ${t.getMessage}", t)
+
+      case ProcessingFailure(q, _, e, p, t) =>
+        log.debug(s"PROCESSING_FAILURE: `$q` after ${(e + p).toMillis} ms (${e.toMillis} ms exec, ${p.toMillis} ms proc (failed)), detail: ${t.getMessage}", t)
+    }
+
+  private def error[F[_]: Sync](msg: => String, cause: => Throwable): F[Unit] =
+    Sync[F].delay(log.error(msg, cause))
+
+  private def debug[F[_]: Sync](msg: => String): F[Unit] =
+    Sync[F].delay(log.debug(msg))
+
+  private def trace[F[_]: Sync](msg: => String): F[Unit] =
+    Sync[F].delay(log.trace(msg))
+
+  private def logChunkSize[F[_]: Sync](c: Chunk[Byte]): F[Unit] =
+    trace[F](s"Sending ${c.size} bytes")
+
+  private def recordChunks[F[_]: Sync](total: Ref[F, Long])(c: Chunk[Byte]): F[Unit] =
+    total.update(_ + c.size) >> logChunkSize[F](c)
+
+  private def copyToTable(table: Table, columns: NonEmptyList[TableColumn]): Pipe[CopyManagerIO, Chunk[Byte], Unit] = {
+    val cols =
+      columns
+        .map(c => hygienicIdent(c.name))
+        .intercalate(", ")
+
+    val copyQuery =
+      s"COPY ${hygienicIdent(table)} ($cols) FROM STDIN WITH (FORMAT csv, HEADER FALSE, ENCODING 'UTF8')"
+
+    val logStart = debug[CopyManagerIO](s"BEGIN COPY: `${copyQuery}`")
+
+    val startCopy =
+      Stream.bracketCase(PFCM.copyIn(copyQuery) <* logStart) { (pgci, exitCase) =>
+        PFCM.embed(pgci, exitCase match {
+          case ExitCase.Completed => PFCI.endCopy.void
+          case _ => PFCI.isActive.ifM(PFCI.cancelCopy, PFCI.unit)
+        })
+      }
+
+    in => startCopy flatMap { pgci =>
+      in.map(_.toBytes) evalMap { bs =>
+        PFCM.embed(pgci, PFCI.writeToCopy(bs.values, bs.offset, bs.length))
+      }
+    }
+  }
+
+  private def createTable(table: Table, columns: NonEmptyList[TableColumn]): ConnectionIO[Int] = {
+    val preamble =
+      fr"CREATE TABLE" ++ Fragment.const(hygienicIdent(table))
```

Comment:
BTW we're going to be doing sanitization in the frontend, so that we can show users what the actual table name. We can keep this for now, but we'll need to remove it later.

Reply:
Hm, that seems...unsafe. Like anyone that has access to the SDBE API could perform SQL injection.

Reply:
Basically we would need to reject unhygienic identifiers (and similar), rather than trying to fix them.